InputFormat support for Iceberg

rdsr opened the pull request:
This addresses #170 
The current code is pretty rough. I'm sending to get early feedback on the approach.
Most of the scaffolding is  there.  There's support to allow for parameterized in memory records. All read value functions for data formats like Avro, Parquet and ORC are supported through the `ReadSupport` interface. The same interface also allows to add identity partition columns to the input row. 
I supported parameterized input records because much of the use of the InputFormat api is for supporting  MR based engines like Pig and Hive. 

I need feedback on the `ReadSupport` API and see whether folks are OK with it.  Detailed code feedback can be omitted for now. Once people are ok the high level approach I will clean up the code and add tests. Feedback welcome on testing the input format as well. 

cc @rdblue @teabot @jerryshao @massdosage

This patch is based on the work of @guilload

jerryshao commented:
I think @teabot they're also working on this, not sure how do we collaborate. Maybe we should create a milestone and break down all the tasks, so that we could parallelize all the works and reduce the conflicts.

jerryshao commented on build.gradle:
```diff
@@ -193,6 +193,24 @@ project(':iceberg-hive') {
+      exclude group: 'org.apache.avro', module: 'avro'
+    }
+
+    testCompile project(path: ':iceberg-hive', configuration: 'testArtifacts')
```
Why do we need to depend on `iceberg-hive` module? Not sure if this will potentially introduce circular dependency.

rdsr replied:
This is so that I can rely on `HiveCatalog`. Yes you are right though, we'd have to figure out how to arrange out modules when we have a Hive storage handler

rdblue replied:
Maybe this could just rely on a catalog from the API, and include iceberg-hive for tests. That wouldn't introduce a dependency cycle.

rdsr replied:
We'd need the `org.apache.iceberg.hive.HiveCatalogs#loadCatalog` which is available in iceberg-hive

rdblue replied:
I think we have some options. We could add a config option to the builder to set the catalog, or we could use reflection to call `HiveCatalogs.loadCatalog` if the class is present.

massdosage replied:
I can see a future where we have Hive SerDes and StorageHandlers in the `hive` subproject which in turn will need to depend on the input format(s) here so we need to avoid a circular dependency right? In our [InputFormat](https://github.com/ExpediaGroup/hiveberg/blob/master/src/main/java/com/expediagroup/hiveberg/IcebergInputFormat.java) up to now we've just use HadoopTables and the JobConf to get everything we need.

rdsr replied:
I fixed this. For now, I've adding iceberg-hive as a test dep

rdsr replied:
Maybe iceberg-hive as a test dep can also cause recursive deps. If that's the case we can simply test with `HadoopCatalog` instead and remove iceberg-hive altogether

rdsr replied:
I've removed iceberg-hive altogether. Relying on HadoopCatalog to test the Catalog plugin

openinx commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,345 @@
+      }
+    }
+
+    private Iterator<T> open(FileScanTask currentTask, Schema readSchema) {
```
I read the pig & spark source reader, seems we could abstract the FileScanTask open(..) method with readSchema,  so that all the compute engine can share that part of code,  maybe we can defined it as: 
```
Iterator<T> open(FileScanTask currentTask, Schema readSchema, ReadSupport<T> readSupport).
```
I think it could be a file-level abstraction ..

jerryshao replied:
I also have the same feelings, with more framework supported, there could be a bunch of duplicated codes. It would be better to have an abstract layer which could be adapted to all the frameworks.

rdsr replied:
Pig will eventually rely on this InputFormat, but I agree generally, if we can abstract away common code, it would be great.

rdblue replied:
Yeah, there may be a good way to extract common code. For now, let's do that in a separate PR though to keep this focused on the MR support.

rdsr commented:
> I think @teabot they're also working on this, not sure how do we collaborate. Maybe we should create a milestone and break down all the tasks, so that we could parallelize all the works and reduce the conflicts.

+1. I'll create those tasks.

massdosage commented:
OK, so the `InputFormat` we've been working on (https://github.com/ExpediaGroup/hiveberg/blob/master/src/main/java/com/expediagroup/hiveberg/IcebergInputFormat.java) implements the `org.apache.hadoop.mapred.InputFormat` API. I see the one above implements the `org.apache.hadoop.mapreduce.InputFormat` API. Which version of Hive is the latter targetting? I'm pretty sure it won't work on Hive 2.x as we tried that first and then moved to the `mapred` API which we can confirm is working for a simple read path (and we also have a Hive unit test for this at https://github.com/ExpediaGroup/hiveberg/blob/master/src/test/java/com/expediagroup/hiveberg/TestIcebergInputFormat.java). See https://stackoverflow.com/questions/33235199/hadoop-hive-serde-input-format-must-implement-inputformat for some more information on the two APIs.

One possible approach here would be that we move our code from the "hiveberg" repo back into a PR against Iceberg but we would need someone from the Iceberg project to help us with the dependency and version hell that this leads to (specifically Hive, Guava and Avro version conflicts). We could then have a `mapred` input format and a `mapreduce` input format and at that point refactor out the commonality from them.

The InputFormat is also only one part of the picture, we're also intending to add a SerDe (see https://github.com/ExpediaGroup/hiveberg/blob/master/src/main/java/com/expediagroup/hiveberg/IcebergSerDe.java) and related classes and ultimately potentially bundle this all into a Hive StorageHandler. This and other roadmap we have on mind can be found at https://github.com/ExpediaGroup/hiveberg/issues. 

Again, we would happily move all this work over to here and I think that would make it a lot easier to have visibility on all the issues and split the work to remove duplication etc.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,345 @@
+    TableScan scan = table.newScan();
+    //TODO add caseSensitive, snapshot id etc..
+
+    Expression filterExpression = SerializationUtil.deserializeFromBase64(conf.get(FILTER_EXPRESSION));
```
Looks like this serialization (java serializable/gzip/base64) is only used for the filter expression. We should consider (in a separate commit) using a parser for Iceberg expressions so that we can convert them to infix strings instead and parse them when needed. That makes the job configuration readable, which is always a good thing.

rdblue replied:
We could also serialize to/from JSON if we wanted something easier than including or building a parser.

rdsr commented:
Thanks @massdosage . 
This implements the MRv2 InputFormat api. Since Pig implements MRv2 and it is the relatively newer api when compared to Mrv1 I chose to implement this.

I think we should be able to add Mrv1 support on top of Mrv2 . I have seen some libraries which provide out of the box support for converting MRv2 to Mrv1. I was unable to find that library now, but I vaguely recollect that maybe twitter had built something around this.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/ReadSupport.java:
```diff
@@ -0,0 +1,50 @@
+ * to the input row
+ * @param <T>
+ */
+public interface ReadSupport<T> {
```
Rather than allowing a user to pass in `ReadSupport`, I think it makes more sense to configure at a higher level. There are good built-in options that don't require exposing classes from Parquet, Avro, and ORC here.

I think the config builder could expose methods to control the in-memory format. By default, it would use Iceberg generic records. Optionally, we could expose Pig's in-memory types and Hive's in-memory types. Maybe Avro as well. Those seem like reasonable options for an `InputFormat` given that users will probably not be customizing their in-memory classes if they are still using the MR read interface. At this point, I would expect the input format to be primarily used for integration with Pig and Hive.

rdsr replied:
I think that makes sense. We either allow Pig's `Tuples`, Hive's `ObjectInspectors` or default Iceberg `GenericRecords`

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,345 @@
+      conf.set(TABLE_SCHEMA, SchemaParser.toJson(table.schema()));
+    }
+
+    public ConfBuilder filterExpression(Expression expression) throws IOException {
```
No need for "Expression" in this API. How about just `filter`?

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,345 @@
+      conf.set(TABLE_SCHEMA, SchemaParser.toJson(table.schema()));
+    }
+
+    public ConfBuilder filterExpression(Expression expression) throws IOException {
```
I'd rather not throw `IOException` here. Can we update `serializeToBase64` to not throw it?

rdsr replied:
+1 .  This was mainly the result of copy pasting error.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,345 @@
+  public static class ConfBuilder {
+    private final Configuration conf;
+
+    public ConfBuilder(Configuration conf, String path, Class<? extends ReadSupport<?>> readSupportClass) {
```
If we are going with the builder pattern, then let's set the table/path and in-memory data model in builder methods instead of here. I prefer not to minimize constructor args with builders by removing anything that might not need to be set (e.g. use Iceberg generics by default).

rdsr replied:
in-memory data model +1.
For table path, which is mandatory, does it make more sense in the constructor? Then users cannot omit it.  Granted that if they omit it, the code will just fail, but I general add required args as const. params and options and methods to the builder.

rdsr replied:
Decided to set everything as builder methods except configuration, similar to Kite

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,345 @@
+import org.slf4j.LoggerFactory;
+
+
+public class IcebergInputFormat<T> extends InputFormat<Void, T> {
```
Can we add documentation?

rdsr replied:
Added.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,345 @@
+    }
+
+    //TODO: other options split-size, snapshotid etc..
+    public Configuration updatedConf() {
```
I'm not sure that this is needed. Unless this builder makes a copy of the `Configuration`, it is configuring the one passed in as methods are called. I think that's probably fine.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,345 @@
+    return new IcebergRecordReader();
+  }
+
+  public static ConfBuilder updateConf(
```
I like the builder pattern here. I used something similar for Kite's `InputFormat`.

You might consider how that builder worked. The static methods there would [accept a `Job` when creating the builder](https://github.com/kite-sdk/kite/blob/master/kite-data/kite-data-mapreduce/src/main/java/org/kitesdk/data/mapreduce/DatasetKeyInputFormat.java#L191-L203). That way, there was no need for the user to also set the input format class. It was also not necessary to build the resulting `Configuration` and do something with it:

```java
  /**
   * Configures the {@code Job} to use the {@code IcebergInputFormat} and
   * returns a helper to add further configuration.
   *
   * @param job the {@code Job} to configure
   */
  public static ConfigBuilder configure(Job job) {
    job.setInputFormatClass(IcebergInputFormat.class);
    return new ConfigBuilder(job.getConfiguration());
  }
```

rdsr replied:
This looks great. Will use

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,345 @@
+    final String schemaStr = conf.get(READ_SCHEMA);
+    if (schemaStr != null) {
+      // Not sure if this is having any effect?
+      scan.project(SchemaParser.fromJson(schemaStr));
```
The base scan will combine the requested columns with the ones used by row filters to produce the required schema. If you don't intend to use that, then it isn't needed. But you could add the schema to the splits to know what to project.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,345 @@
+            .impl(readSupportClassName)
+            .<ReadSupport<T>>buildChecked()
+            .newInstance();
+      } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
```
You might want to use `DynConstructors` and `build` instead of `buildChecked`. That will get you better error messages and some of this error handling for free.

rdsr commented:
Any suggestions on how we support the mapred API? 
We can go the route of https://github.com/twitter/elephant-bird/blob/master/core/src/main/java/com/twitter/elephantbird/mapred/input/DeprecatedInputFormatWrapper.java.
Any other ideas?
@rdblue , @massdosage, @jerryshao

rdsr commented:
Addressed some review comments.
Specifically around having a higher level way of specifying in memory data model.
If folks can eye ball the PR , that will be great. Three main things I need feedback on
1. In Memory data model approach
2. Logic for handling identity partition columns
3. How we do address mapred inputformat as described above.

rdsr commented:
Once folks are ok with the approach. I'll start writing tests

jerryshao commented:
>Any suggestions on how we support the mapred API?
We can go the route of https://github.com/twitter/elephant-bird/blob/master/core/src/main/java/com/twitter/elephantbird/mapred/input/DeprecatedInputFormatWrapper.java.
Any other ideas?

I'm not so if there're other wrappers, but I think this should be fine.

rdsr commented:
> I'm not so if there're other wrappers, but I think this should be fine.
I did see long time back a better written converter from MRv2 -> Mrv1 I couldn't remember the name of the library now.  
I guess if we go the route of the converter, we'd have to copy paste that code instead of adding that dependency which could bring in a lot of unneeded artifacts.

rdsr commented:
> Any suggestions on how we support the mapred API?
We can go the route of https://github.com/twitter/elephant-bird/blob/master/core/src/main/java/com/twitter/elephantbird/mapred/input/DeprecatedInputFormatWrapper.java.
Any other ideas?
@rdblue , @massdosage, @jerryshao

I'll create a ticket for this

massdosage commented:
I've raised a [PR](https://github.com/rdsr/incubator-iceberg/pull/1) against @rdsr's branch which could be a first step at having both InputFormats, we could then look into factoring the commonality out (possibly using something like ElephantBird). Alternatively we raise a separate PR for the `mapreduce` InputFormat and do the refactoring of commonality later. 

It would be good to decide on who is going to do what and in what order. For now we are working on the Hive Serde and looking into improving the Hive read path (other file formats, partition predicate and projected column pushdown etc.)

rdblue commented:
Do we need to support the `mapred` format? If the intent is to support Pig, Hive, and MR applications then isn't a `mapreduce` format sufficient?

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,494 @@
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext newContext) {
+      Configuration conf = newContext.getConfiguration();
+      CombinedScanTask task = ((IcebergSplit) split).task;
```
We should note in documentation that this `InputFormat` must perform its own split planning and doesn't accept `FileSplit` instances. I think it's fairly common to create and pass file splits, so we may just want to accept them and treat them as single-file tasks.

rdsr replied:
If the splits are only created using Iceberg apis, I'd imagine we will only be getting IcebergSpits , no?

rdblue replied:
Yes, if splits are produced using `getSplits` then we'd be fine. I seem to remember some problem doing that with Hive for Parquet, though. We ended up using regular `FileSplit` there.

rdsr commented:
@rdblue  HiveStorageHandler https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java only supports the mapred API

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,494 @@
+
+    splits = Lists.newArrayList();
+    try (CloseableIterable<CombinedScanTask> tasksIterable = scan.planTasks()) {
+      tasksIterable.forEach(task -> splits.add(new IcebergSplit(conf, task)));
```
I just realized that we also need to assert that the residual for each file in the task is either `null` or `Expressions.alwaysTrue`. That will guarantee that the filter expression is completely satisfied and no additional rows will be returned.

Alternatively, we could run the residual to filter every row in the reader, but it's probably simpler for now to open that as a follow-up issue.

rdsr replied:
I think Pig and Hive could apply the residuals on their own if we return the pushed down filters as is, like we do in Spark.  For MR standalone, we can run the residual like we do for IcebergGenerics for now I do as you mentioned but only for MR standalone.

rdblue replied:
I like that this applies the residuals for generics, but I think we need to be a bit safer for the others. I think we should make a config option to enable the current behavior with Pig and Hive object models. If it is set, then Pig and Hive are expected to apply the final filter on their own. If not, then we assert that the residual is null or alwaysTrue. I think it's much safer to make skipping residuals opt-in like that.

rdsr replied:
Fixed.

rdsr replied:
Added follow up for residual evalution for Iceberg generics - https://github.com/apache/incubator-iceberg/issues/866

rdsr commented:
@massdosage Thanks for the PR. I'll look into it.
If it is OK with you, I  suggest the the following break down
You can implement the Hive StorageHandler, Serde and mapred inputformat [lets discuss this here what is the best way to do this]. I'll help with the reviews.
I can continue working on the mapreduce API 
Let me know if it sounds good to you and if anything else is required.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,494 @@
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          // TODO implement adding partition columns to records for Pig and Hive
```
I think we need to update how we handle partition columns in general. #585 points out that the current row-join approach doesn't work for nested columns. We'll need to update row materialization to fix it. I can work on that sometime soon because we also need it for metadata columns in Spark (in particular, row ordinal in a file).

FYI @aokolnychyi.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,494 @@
+      return this;
+    }
+
+    public ConfigBuilder inMemoryDataModel(InMemoryDataModel inMemoryDataModel) {
```
What about changing this to `usePigTuples()` and `useHiveRows()`? That eliminates the need for a public enum and I think is easy to read.

rdsr replied:
+1. I like yours better!

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,494 @@
+    }
+  }
+
+  private static Table getTable(Configuration conf) {
```
In Spark, this is an instance method so that user can override it and use their own catalog. I think that would be a good idea here as well:

```java
public class CustomCatalogIcebergInputFormat extends IcebergInputFormat {
  @Override
  public Table getTable(String tablePath) {
    CustomCatalog.get().loadTable(conf.get(tablePath));
  }
}
```

In that example, I'm assuming that we'd have a `final getTable(Configuration)` method that pulls `TABLE_PATH` out of the config and passes it to `getTable`.

rdsr replied:
I think this is slightly related to the HiveCatalog discussion above where we want to load HiveCatalog through reflection if HiveCatalog is on the classpath [since we do not want to add the iceberg-hive dependency]

The problem here is that this `getTable` method needs to be static since it is being called from the `configBuilder`.

To solve both these issues - not relying on hive-deps at compile time and ability to plug in a custom catalog, can we define a `CatalogFactory` which allows us to load our own custom catalogs. The `CatalogFactory` can be passed as a config property. Since this factory is an interface, it should be easier to instantiate [through reflection] and then we can load our custom catalog.

What do u think?

rdblue replied:
Yeah, I guess that's not a good way to inject a catalog. In addition, the configuration methods sets up the `IcebergInputFormat` and not a user's subclass.

I think the factory idea would work, but it's more of a `Supplier<Catalog>` since there is no configuration. Like this?

```java
  ConfigBuilder catalogSupplier(Class<? extends Supplier<Catalog>> supplierClass)) {
    this.supplierClass = supplierClass;
    return this;
  }
```

rdsr replied:
Supplier can work, but how do we pass the Configuration to the  Catalog  to instantiate. Supplier is a function with no args. Maybe a `Function<Configuration,Catalog>` works better?

rdblue replied:
You're right, we probably want to use `Function` instead.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,494 @@
+        return locations;
+      }
+
+      Set<String> locationSets = Sets.newHashSet();
```
Is this duplicated? Maybe we should have a locality util class.

rdblue commented on mr/src/test/java/org/apache/iceberg/mr/TestIcebergInputFormat.java:
```diff
@@ -0,0 +1,133 @@
+        icebergInputFormat.createRecordReader(splits.get(0), context);
+    recordReader.initialize(splits.get(0), context);
+    while (recordReader.nextKeyValue()) {
+      System.out.println(recordReader.getCurrentValue());
```
We should make this an assertion before committing this PR.

rdsr replied:
Tests are not done well. I plan to write better tests before publishing this PR

rdsr replied:
Fixed.

rdblue commented:
>  HiveStorageHandler only supports the mapred API

Sounds good then.

rdblue commented:
> Three main things I need feedback on
> 1. In Memory data model approach
> 2. Logic for handling identity partition columns
> 3. How we do address mapred inputformat as described above.

1. I like how it works now.
2. I commented above; I think we need to configure the readers with a map from ID to value and have them insert identity-partition values at the right place. This is a larger issue than just this `InputFormat` because it also affects Spark (#585).
3. The `mapred` and `mapreduce` classes aren't all that different. I'd prefer to either wrap or make common utilities used by the two in Iceberg, rather than relying on a dependency that we have to worry about.

rdsr commented:
Thanks @rdblue 
For identity partitioning, I'll keep the current code and we can address it later through #585 
For mapred I think once the mapreduce API gets committed we can see how best to refactor to support mapred API

jerryshao commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,494 @@
+  static final String SPLIT_SIZE = "iceberg.mr.split.size";
+  static final String TABLE_PATH = "iceberg.mr.table.path";
+  static final String TABLE_SCHEMA = "iceberg.mr.table.schema";
+  private static final String LOCALITY = "iceberg.mr.locality";
```
Maybe we'd also consider adding incremental support here, like what current Spark `Reader` did.

rdblue replied:
Let's do that in a follow-up.

jerryshao commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,494 @@
+    } else {
+      Catalog catalog = HiveCatalogs.loadCatalog(conf);
+      TableIdentifier tableIdentifier = TableIdentifier.parse(path);
+      return catalog.loadTable(tableIdentifier);
```
For a simple MR application, do we need to involve `HiveCatalog`?

It seems strange to me that a simple MR application requires to communicate with Hive. I was thinking that if we could have an abstract catalog layer here, or only support Hadoop tables in MR as default implementation. Besides let Hive, Pig and others to choose a specific implementation, because each engine may have their own preference, and we don't have to bind to Hive or others here.

rdblue replied:
I like this idea, but I don't think it works. The problem is that `getSplits` isn't called in the client/driver process. It's called in the setup task that needs to instantiate the `InputFormat` with the no-arg constructor. So everything needs to be passed through the `Configuration`.

Tables themselves aren't serializable, so we can't pass a table or a scan that way. We need to be able to load the table in the setup task, configure the scan, and get splits there.

massdosage commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,494 @@
+/*
```
Can I suggest putting this and the related classes into a `org.apache.iceberg.mr.mapreduce` package? So we can then add the `org.apache.iceberg.mr.mapred` implementation alongside it and the package names match the Hadoop API package names?

rdsr replied:
@massdosage . I'll make this change once all other comments are addressed. I think making this change before can trip up github and it may consider mr.InputFormat and mr.mapreduce.InputFormat as two different files.

massdosage replied:
OK, sure.

rdsr replied:
Fixed.

rdsr commented on mr/src/test/java/org/apache/iceberg/mr/TestIcebergInputFormat.java:
```diff
@@ -0,0 +1,260 @@
+  public static Object[][] parameters() {
+    return new Object[][]{
+        new Object[]{"parquet"},
+        new Object[]{"avro"}
```
Will add orc once ORC patches are committed which read/writer Icebreg generics

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,525 @@
+
+    private Iterator<T> open(FileScanTask currentTask, Schema readSchema) {
+      DataFile file = currentTask.file();
+      // TODO should we somehow make use of FileIO to create inputFile?
```
Yes. `FileIO` is `Serializable` and should be used to create input files. We should also use the encryption manager to be correct (could open a follow-up).

rdsr replied:
I'll create a follow up for this if that's OK. Should this be serialized as base64 and added to job conf. I think job configuration generally has a default size limit of 5MB

rdblue replied:
Yeah, sounds good.

rdsr replied:
Created a follow up #865

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,525 @@
+      if (filter == null || filter.equals(Expressions.alwaysTrue())) {
+        return iterable;
+      } else {
+        throw new UnsupportedOperationException(
```
Is it possible to throw this in `getSplits` so it isn't an error that may happen 95% of the way through a job?

rdsr replied:
Fixed

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,525 @@
+      scan = scan.asOfTime(asOfTime);
+    }
+    long splitSize = conf.getLong(SPLIT_SIZE, -1);
+    if (splitSize != -1) {
```
Minor: could we use > 0 instead of != -1?

rdblue commented on build.gradle:
```diff
@@ -219,6 +219,58 @@ project(':iceberg-hive') {
+    // that's really old. We use the core classifier to be able to override our guava
+    // version. Luckily, hive-exec seems to work okay so far with this version of guava
+    // See: https://github.com/apache/hive/blob/master/ql/pom.xml#L911 for more context.
+    testCompile("org.apache.hive:hive-exec::core") {
```
Are Hive dependencies still needed for `testCompile`?

To test this with HMS, we could make it a test dependency of `iceberg-hive` and test out the catalog function.

rdsr replied:
I'l also worried that iceberg-hive in test scope can create a circular dep. Maybe we can just test with `HadoopCatalog` ?

rdblue replied:
Why would it create a circular dependency? This doesn't need to depend on iceberg-hive anymore, right?

rdsr replied:
We'd need to depend on iceberg-hive in test scope so that we can use HiveCatalog

rdblue replied:
What about putting the test case for HiveCatalog in iceberg-hive using the catalog function? Then we get a test that uses the catalog plugin system and we don't have to include Hive here.

rdsr replied:
I've remove the iceberg-hive and related deps from iceberg-mr module and the plugin system is being tested using `HadoopCatalog`.  I can still add the test for HiveCatalog in the iceberg-hive module. What do you think?

massdosage commented on core/src/main/java/org/apache/iceberg/hadoop/Util.java:
```diff
@@ -36,4 +47,21 @@ public static FileSystem getFs(Path path, Configuration conf) {
+          locationSets.addAll(Arrays.asList(b.getHosts()));
+        }
+      } catch (IOException ioe) {
+        LOG.warn("Failed to get block locations for path {}", path, ioe);
```
Is it OK if this happens? We don't want to rethrow this?

rdsr replied:
For now, I've kept it as it was for Spark. Seem like locality is not really needed for cloud stores and it is an optimization for HDFS. I can throw an exception and then handle it across Spark and MR if you guys think this is necessary..

jerryshao replied:
Locality preference is just a nice-to-have thing, to throw an exception for this seems not so necessary.

massdosage replied:
OK, makes sense.

rdblue replied:
Yes, locality is optional and not being able to get it should not cause a job to fail.

jerryshao commented on mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java:
```diff
@@ -0,0 +1,567 @@
+    public String[] getLocations() {
+      boolean localityPreferred = conf.getBoolean(LOCALITY, false);
+      if (!localityPreferred) {
+        return ANYWHERE;
```
Does MR treat "*" as `ANYWHERE`?

rdsr replied:
I copied this from IcebergPigInputFormat. I'll double check

jerryshao commented on mr/src/test/java/org/apache/iceberg/mr/TestIcebergInputFormat.java:
```diff
@@ -0,0 +1,242 @@
+
+
+@RunWith(Parameterized.class)
+public class TestIcebergInputFormat {
```
Maybe we should add more tests to cover some of the configurations defined above.

rdblue replied:
We also need tests that use the MR framework, not just tests that independently use the input format API.

You can use [this s3committer test](https://github.com/rdblue/s3committer/blob/master/src/test/java/com/netflix/bdp/s3/TestMRJob.java#L50) as an example. That test uses a mock to test s3 that isn't needed, but it shows how to bring up a test DFS and MR cluster.

rdsr replied:
I tried writing a test with DFS and MR cluster. I'm facing classpath issues particularly with Guava. Any suggestions on how to go about it? I was think maybe we can have an integration testing module per  framework. For example for MR we'd have `mr-runtime` module which would be shaded. Now to test MR we could have a `mr-runtime-test` module which can setup DFS and MR clusters. Since mr runtime would be shaded completely we can write these integration tests.

I've added more unit test though.

rdsr replied:
This is the actual error
```
java.lang.NoSuchMethodError: com.google.common.base.Objects.toStringHelper(Ljava/lang/Object;)Lcom/google/common/base/Objects$ToStringHelper;

	at org.apache.hadoop.metrics2.lib.MetricsRegistry.toString(MetricsRegistry.java:406)
	at java.lang.String.valueOf(String.java:2994)
	at java.lang.StringBuilder.append(StringBuilder.java:131)
	at org.apache.hadoop.ipc.metrics.RpcMetrics.<init>(RpcMetrics.java:74)
	at org.apache.hadoop.ipc.metrics.RpcMetrics.create(RpcMetrics.java:80)
	at org.apache.hadoop.ipc.Server.<init>(Server.java:2218)
	at org.apache.hadoop.ipc.RPC$Server.<init>(RPC.java:951)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.<init>(ProtobufRpcEngine.java:534)
	at org.apache.hadoop.ipc.ProtobufRpcEngine.getServer(ProtobufRpcEngine.java:509)
	at org.apache.hadoop.ipc.RPC$Builder.build(RPC.java:796)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.<init>(NameNodeRpcServer.java:351)
	at org.apache.hadoop.hdfs.server.namenode.NameNode.createRpcServer(NameNode.java:674)
```
I've the MR framework tested in my own repo https://github.com/rdsr/incubator-iceberg/blob/mr_generic_job/mr/src/test/java/org/apache/iceberg/mr/TestMRJob.java

jerryshao replied:
Yes, this could be a problem. I agree that we should have a mr-runtime module to shade some conflict dependencies.

rdblue replied:
Do you know what version of Guava is getting pulled in? The Iceberg version?

rdsr replied:
My guess would be yes as `toStringHelper` is not defined in our version of guava `28.0-jre`

massdosage replied:
We've had similar issues in our branch where we are trying to get the Hive InputFormat to work. Hive 2.3.6 requires Guava 11.0.2, if a newer Guava version is on the classpath Hive is unable to use the InputFormat due to exceptions similar to the one above.  So we have to remove Guava as an exposed dependency from all Iceberg artifacts which appear on the Hive classpath. The only way we've managed to get it to work is by doing the following:

* Alter every Iceberg module that uses Guava to shade and relocate it (which IMHO is a good thing to do anyway so external users of Iceberg can use their own versions of Guava).
* Depend on the shaded version of these modules from iceberg-mr.
* Remove Guava from `versions.props` so that different subprojects can depend on different versions of it.
* The Guava version that then gets used in iceberg-mr is the transitive one from Hive 2.3.6 (in this case) which is Guava 11.0.2.

You can see these changes here: https://github.com/ExpediaGroup/incubator-iceberg/blob/078a06ddd78d08648127d8b2e8dc41e0febf7f49/build.gradle We're not Gradle experts so hopefully there is an easier way to do all of this but I think the general steps outlined above will still be required.

Ultimately I think this issue is going to have to be solved for both InputFormats so any changes that would allow different versions of Guava to be used would be great.

rdblue replied:
I think it's reasonable to build a mr-runtime library. It's really too bad that this requires one, since it is difficult to get the licensing right for a binary Jar like that and it's more maintenance to have 2 more modules. Hopefully we can reuse some of the work from Spark's runtime Jar.

jerryshao commented on core/src/main/java/org/apache/iceberg/hadoop/Util.java:
```diff
@@ -19,13 +19,24 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
```
one more empty line.

rdblue commented on mr/src/test/java/org/apache/iceberg/mr/TestIcebergInputFormat.java:
```diff
@@ -0,0 +1,242 @@
+import org.junit.runners.Parameterized;
+
+import static org.apache.iceberg.types.Types.NestedField.required;
+
```
Nit: extra newline.

rdblue commented:
@rdsr, the implementation looks good to me. Just need more tests, like @jerryshao pointed out.

rdsr commented:
@rdblue added more unit tests. Please have a look.  Had classpath issues with mini cluster. I've explained the problem above

rdsr commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -346,14 +348,15 @@ public void close() throws IOException {
       boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
+
       if (hasJoinedPartitionColumns) {
-        readSchema = TypeUtil.selectNot(tableSchema, idColumns);
```
Had to change this logic slightly so that whatever schema the user projected, that is what is returned [we drop identity partition columns not projected by the user]

jerryshao commented on mr/src/main/java/org/apache/iceberg/mr/SerializationUtil.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Base64;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+
+
```
additional line.

jerryshao commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
```
Also here.

jerryshao commented:
LGTM, just some minor comments.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+    public ConfigBuilder catalogFunc(Class<? extends Function<Configuration, Catalog>> catalogFuncClass) {
+      Preconditions.checkState(
+          conf.get(TABLE_PATH) == null,
+          "Please provide custom catalog before specifying the table to read from");
```
Why is this necessary? Can't the work be done in either method when there is enough information to load the table?

rdsr replied:
You are right. We don't need this.  If `readFrom` is called before calling `catalogFunc` we through a well defined error saying no custom catalog provided.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+    //TODO remove the check on dataModel once we start supporting
+    // residual evaluation for Iceberg Generics in InputFormat
+    InMemoryDataModel dataModel = conf.getEnum(IN_MEMORY_DATA_MODEL, InMemoryDataModel.DEFAULT);
+    if (dataModel == InMemoryDataModel.DEFAULT || !platformAppliesFilter) {
```
I think that if the job configured `plaformAppliesFilter` then it is okay to return records even if the model is generics.

rdsr replied:
fixed.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+  private enum InMemoryDataModel {
+    PIG,
+    HIVE,
+    DEFAULT // Default data model is of Iceberg Generics
```
I think this should be `GENERIC` instead of `DEFAULT`. Defaults depend on context, but this enum doesn't need to encode what some other part of the code will do by default. It just needs to encode the data model.

rdsr replied:
fixed

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+     * filter is not completely satisfied. Note: This
+     * does not apply to standalone MR application
+     */
+    public ConfigBuilder platformAppliesFilterResiduals() {
```
I find this a little hard to reason about and I think a different name could make it more straight-forward. Rather than configuring "someone will do task" it is easier for the caller and this code to phrase it as "do task" or "skip task". That also avoids the need to refer to the "platform" because users won't know what that is and we have to explain it.

What about changing this to `skipResidualFiltering()`? That is clear that the `InputFormat` is not responsible for filtering, but doesn't define who or what is responsible. It is also clearly dangerous, even if you don't know what a residual is.

rdsr replied:
fixed

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+     * should call this api. Otherwise the current
+     * api will throw an exception if the passed in
+     * filter is not completely satisfied. Note: This
+     * does not apply to standalone MR application
```
Nit: the line wrapping here is strange.

rdsr replied:
fixed

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+  }
+
+  private static void checkResiduals(Configuration conf, CombinedScanTask task) {
+    boolean platformAppliesFilter = conf.getBoolean(PLATFORM_APPLIES_FILTER_RESIDUALS, false);
```
Minor: this calls `getBoolean` and `getEnum` for every split. It would be nice to get those outside the loop and pass them in.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+      task.files().forEach(fileScanTask -> {
+        Expression residual = fileScanTask.residual();
+        if (residual != null && !residual.equals(Expressions.alwaysTrue())) {
+          throw new RuntimeException(
```
How about `UnsupportedOperationException` since this will eventually be implemented?

rdsr replied:
Well, this would only be implemented for GENERIC data model, not for Pig and Hive. Thoughts on keeping the same exception or do u think we should implement residual eval for Pig and Hive data models?

rdblue replied:
My rationale was that we would accept a patch to add residual filtering for Hive and Pig, so why not consider it a future feature? Should someone choose to contribute it I think it would be valuable.

rdsr replied:
OK that make sense

rdsr replied:
Kept as `UnsupportedException`

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+    private InMemoryDataModel inMemoryDataModel;
+    private Closeable currentCloseable;
+    private boolean reuseContainers;
+    private boolean caseSensitive;
```
Nit: I prefer separating state that won't change (`caseSensitive`, `reuseContainers`) from state that does change (`currentRow`, `currentCloseable`), but those are mixed together here.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+
+    @Override
+    public float getProgress() {
+      return context.getProgress();
```
Maybe leave a TODO item to improve this.

rdsr replied:
@rdblue , any thoughts on what we need to improve here?

rdblue replied:
Just that we could give a more accurate progress based on records read from the file. I don't think that `context.getProgress` has enough information to give an accurate progress value.

This isn't that easy, since we don't know how much of the input split has been processed and we are pushing filters into Parquet and ORC. But we do know when a file is opened and could count the number of rows returned, so we can estimate. And we could also add a row count to the readers so that we can get an accurate count of rows that have been either returned or filtered out.

rdsr replied:
Added

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+    private Iterator<T> open(FileScanTask currentTask, Schema readSchema) {
+      DataFile file = currentTask.file();
+      // TODO should we somehow make use of FileIO to create inputFile?
+      InputFile inputFile = HadoopInputFile.fromLocation(file.path(), context.getConfiguration());
```
In the future, we should update HadoopFileIO to use a transient `Configuration` and to keep a set of properties that it extracts from its original configuration. Then we can serialize HadoopFileIO more easily, and rebuild a working IO config on tasks.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+    String catalogFuncClass = conf.get(CATALOG);
+    if (catalogFuncClass != null) {
+      Function<Configuration, Catalog> catalogFunc
+          = (Function<Configuration, Catalog>)
```
Nit: `=` should be on the previous line.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+      Catalog catalog = catalogFunc.apply(conf);
+      TableIdentifier tableIdentifier = TableIdentifier.parse(path);
+      return catalog.loadTable(tableIdentifier);
+    } else if (path.contains("/")) {
```
I think this should be the first branch of the `if` because it is the easiest and is less likely to fail.

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+    private CloseableIterable<T> newAvroIterable(InputFile inputFile, FileScanTask task, Schema readSchema) {
+      Avro.ReadBuilder avroReadBuilder = Avro.read(inputFile)
+                                             .project(readSchema)
+                                             .split(task.start(), task.length());
```
Nit: typical style is to use 2 indents / 4 spaces from the next line, not to indent to the same level as the last function call.

rdblue commented on mr/src/test/java/org/apache/iceberg/mr/mapreduce/TestIcebergInputFormat.java:
```diff
@@ -0,0 +1,382 @@
+  public static Object[][] parameters() {
+    return new Object[][]{
+        new Object[]{"parquet"},
+        new Object[]{"avro"}
```
Why not ORC?

rdblue replied:
(Now that generics for ORC have been added)

rdsr replied:
Added!

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java:
```diff
@@ -0,0 +1,571 @@
+    }
+
+    private static Record genericRecordWithPartitionsColumns(
+        Record record, Schema identityPartitionSchema, PartitionSpec spec, StructLike partition) {
```
There are tests for this in Spark here: https://github.com/apache/incubator-iceberg/commit/6cafdab76864685d5489ba03cd041865b5b57f01

You might consider adding some similar tests.

rdsr replied:
Added similar tests. Had to change the logic of handling identity columns.

rdblue commented:
@rdsr, this looks almost ready to me. The main thing is to add the MR tests that use a mini-DFS and mini-MR cluster.

rdsr commented:
@rdblue . Regarding mini-DFS and mini-MR I'm facing classpath issues as explained here https://github.com/apache/incubator-iceberg/pull/843#discussion_r402075912 . I've also outlined an approach on how to add a test using iceberg-mr-runtime and iceberg-mr-runtime-test. Let me know if the mini dfs can be addressed as a separate PR

rdblue commented:
Looks good! I merged this. Thanks for the contributions and reviews, everyone!
